### PR TITLE
Stronger typing

### PR DESCRIPTION
**Pull request by jvanderleeuw**
Changes were made to agpypeline, specifically to entrypoint.py and environment.py based on KyClark's comments: https://github.com/kyclark/transformer-canopycover/blob/kyc_testing/REVIEW.adoc 

Argparse now does validation for input files (metadata and images), whereas before this validation was being done within agpypeline itself, and the tests in test_entrypoint.py were changed in order to reflect this change. 

For metadata, files are now in the readable text format and so might be opened initially, but then these are closed if not already closed in load_metadata_files. Specifying bad metadata or image files should now be handled primarily by argparse.

One peculiarity is that in order for the algorithms to work, file_list must be the first argument since it is a positional argument, and if it is specified after --metadata, there will be errors since metadata is required and can have more than one file

The new working version of agpypeline used for this branch is located at https://test.pypi.org/project/agpypeline/0.0.88/

https://app.zenhub.com/workspaces/ua-ag-data-science-5a57a3198339f11ba1c85775/issues/agpipeline/issues-and-projects/185 was also done on this branch and corresponding tests added

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 setuptools.setup(
     name="agpypeline",
-    version="0.0.44",
+    version="0.0.88",
```
Comment (Chris-Schnaufer) [inline setup.py:11]:
```suggestion
    version="0.0.47",
```

File: agpypeline/entrypoint.py [hunk 2]
```diff
@@ -354,9 +358,10 @@ def add_parameters(parser: argparse.ArgumentParser, algorithm_instance: Algorith
     parser.add_argument('--result', nargs='?', default='all',
                         help='Direct the result of a run to one or more of (all is default): "all,file,print"')
 
-    parser.add_argument('--metadata', type=str, action='append', help='The path to the source metadata')
+    parser.add_argument('-m', '--metadata', type=argparse.FileType('rt'), nargs='+',
+                        required=True, help='The path to the source metadata')
```
Comment (Chris-Schnaufer) [inline agpypeline/entrypoint.py:362]:
```suggestion
    parser.add_argument('-m', '--metadata', type=argparse.FileType('rt'), action='append',
                        help='The path to the source metadata')
```

File: agpypeline/entrypoint.py [hunk 3]
```diff
@@ -368,7 +373,8 @@ def add_parameters(parser: argparse.ArgumentParser, algorithm_instance: Algorith
         algorithm_instance.add_parameters(parser)
 
     # Assume the rest of the arguments are the files
-    parser.add_argument('file_list', nargs=argparse.REMAINDER, help='additional files for transformer')
+    parser.add_argument('file_list', nargs='+', type=str,
+                        help='additional files for transformer')
```
Comment (Chris-Schnaufer) [inline agpypeline/entrypoint.py:377]: Why was this change needed?

File: agpypeline/geoimage.py [hunk 4]
```diff
@@ -205,6 +207,52 @@ def create_geotiff(pixels: np.ndarray, gps_bounds: tuple, out_path: str, srid: i
...
+        # GDAL 3 changes axis order: https://github.com/OSGeo/gdal/issues/1546
+        # pylint: disable=no-member
+        dest_spatial.SetAxisMappingStrategy(osgeo.osr.OAMS_TRADITIONAL_GIS_ORDER)
+
+    if dest_spatial.ImportFromEPSG(int(LAT_LON_EPSG_CODE)) != ogr.OGRERR_NONE:
+        msg = "Failed to import EPSG %s for conversion to lat-lon" % str(LAT_LON_EPSG_CODE)
+        logging.error(msg)
+        raise RuntimeError(msg)
+    ref_sys = osr.SpatialReference()
+    ref_sys.ImportFromEPSG(int(epsg))  # an error would have been caught before this point
```
Comment (Chris-Schnaufer) [inline agpypeline/geoimage.py:244]: > an error would have been caught before this point

It's not clear where else this might have been caught. Can you either update the comment, or, (better yet IMO), put the comparison to `ogr.OGERR_NONE` back?

File: agpypeline/geoimage.py [hunk 5]
```diff
@@ -258,8 +307,7 @@ def get_image_bounds(file_path: str, default_epsg: int = None) -> Optional[ogr.G
     ring.AddPoint(bounds[3], bounds[0])  # lower right
     ring.AddPoint(bounds[2], bounds[0])  # lower left
     ring.AddPoint(bounds[2], bounds[1])  # Closing the polygon
-
-    return geometries.polygon_from_ring(ring, int(epsg))
+    return geometries.polygon_from_ring(ring, int(epsg), file_path)
```
Comment (Chris-Schnaufer) [inline agpypeline/geoimage.py:310]: Did we discuss getting rid of `file_path` and handling a problem here?

File: agpypeline/geometries.py [hunk 6]
```diff
@@ -128,15 +133,16 @@ def geometry_to_geojson(geom: ogr.Geometry, alt_coord_type: str = None, alt_coor
                 'code': ref_sys.GetAttrValue("AUTHORITY", 1)
             }
         }
-
     return json.dumps(geom_json)
 
 
-def polygon_from_ring(ring: ogr.Geometry, epsg: int = None) -> Optional[ogr.Geometry]:
+def polygon_from_ring(ring: ogr.Geometry, epsg: int = None, filename: str = None) -> Optional[ogr.Geometry]:
```
Comment (Chris-Schnaufer) [inline agpypeline/geometries.py:139]: I thought we had talked about getting rid of the `filename` parameter? This should return None when there's a problem and let the calling function do what it thinks is correct

File: setup.py (hunk 1 above)
Comment (Chris-Schnaufer) [inline setup.py:11]:
```suggestion
    version="0.0.47",
```

**Review by Chris-Schnaufer — COMMENTED** [review]
Added other comments that need to be addressed.

After the comments are addressed, this branch needs to be merged on the command line with the `develop` branch so that changes which were already accepted don't show up as new changes.

File: agpypeline/entrypoint.py (hunk 3 above)
Comment (jvanderleeuw) [reply agpypeline/entrypoint.py:377]: I based off of the comments from https://github.com/kyclark/transformer-canopycover/blob/kyc_testing/REVIEW.adoc . And then noted (I thin it might have been in an earlier message from slack) that argparse.FileType opens the file and leaves it open, which can cause issues and so changed it to type=str since otherwise a chunk of the agpypeline code would need to be changed. (Let me know if this should be done instead though)

Screen Shot 2021-01-08 at 11 33 53 AM
Screen Shot 2021-01-08 at 11 34 14 AM

Comment (Chris-Schnaufer) [reply agpypeline/entrypoint.py:377]: That argument is supposed to be optional.

File: .github/workflows/testing_checks.yaml [hunk 7]
```diff
@@ -4,6 +4,7 @@ on:
     branches:
       - master
       - develop
+      - stronger_typing
```
Comment (Chris-Schnaufer) [inline .github/workflows/testing_checks.yaml:7]:
```suggestion
```

**Review by Chris-Schnaufer — CHANGES_REQUESTED** [review]
I think this is almost ready to go

**Review by Chris-Schnaufer — APPROVED** [review]
I think this is almost ready to go
